// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-slide-toggle/dynamic-slide-toggle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicSlideToggleComponent = /** @class */ (function () {
    function TdDynamicSlideToggleComponent() {
        this.label = '';
        this.name = '';
        this.hint = '';
        this.required = false;
    }
    TdDynamicSlideToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-slide-toggle',
                    template: "<div class=\"td-dynamic-slide-toggle-wrapper\">\n  <mat-slide-toggle [formControl]=\"control\"\n                    [attr.name]=\"name\"\n                    [required]=\"required\">\n    {{label}}\n  </mat-slide-toggle>\n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                    styles: [""]
                }] }
    ];
    return TdDynamicSlideToggleComponent;
}());
export { TdDynamicSlideToggleComponent };
if (false) {
    /** @type {?} */
    TdDynamicSlideToggleComponent.prototype.control;
    /** @type {?} */
    TdDynamicSlideToggleComponent.prototype.label;
    /** @type {?} */
    TdDynamicSlideToggleComponent.prototype.name;
    /** @type {?} */
    TdDynamicSlideToggleComponent.prototype.hint;
    /** @type {?} */
    TdDynamicSlideToggleComponent.prototype.required;
}
