// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-file-input/dynamic-file-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicFileInputComponent = /** @class */ (function () {
    function TdDynamicFileInputComponent() {
        this.required = undefined;
        this.label = '';
        this.name = '';
        this.hint = '';
        this.errorMessageTemplate = undefined;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TdDynamicFileInputComponent.prototype._handlefileDrop = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.control.setValue(value);
    };
    TdDynamicFileInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-file-input',
                    template: "<div class=\"td-dynamic-file-input-wrapper\">\n  <mat-form-field tdFileDrop\n                  class=\"td-dynamic-file-input-field\"\n                  floatLabel=\"never\"\n                  [disabled]=\"control?.disabled\"\n                  (fileDrop)=\"_handlefileDrop($event)\"\n                  (click)=\"!control?.disabled && fileInput.inputElement.click()\"\n                  (keyup.enter)=\"!control?.disabled && fileInput.inputElement.click()\"\n                  (keyup.delete)=\"fileInput.clear()\"\n                  (keyup.backspace)=\"fileInput.clear()\">\n    <input matInput\n          [value]=\"control?.value?.name\"\n          [placeholder]=\"label\"\n          [attr.name]=\"name\"\n          [disabled]=\"control?.disabled\"\n          readonly/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n  <button mat-icon-button *ngIf=\"control.value\" (click)=\"fileInput.clear()\" (keyup.enter)=\"fileInput.clear()\">\n    <mat-icon>cancel</mat-icon>\n  </button>\n  <td-file-input class=\"td-file-input\"\n                  #fileInput\n                  [formControl]=\"control\">\n    <mat-icon>folder</mat-icon>\n    <span>{{ label }}</span>\n  </td-file-input>\n</div>",
                    styles: [".td-dynamic-file-input-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-file-input-wrapper .td-dynamic-file-input-field{-ms-flex:1;flex:1;box-sizing:border-box}.td-file-input{margin-left:10px}"]
                }] }
    ];
    return TdDynamicFileInputComponent;
}());
export { TdDynamicFileInputComponent };
if (false) {
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.control;
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.required;
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.label;
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.name;
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.hint;
    /** @type {?} */
    TdDynamicFileInputComponent.prototype.errorMessageTemplate;
}
