// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-input/dynamic-input.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicInputComponent = /** @class */ (function () {
    function TdDynamicInputComponent() {
        this.label = '';
        this.hint = '';
        this.type = undefined;
        this.required = undefined;
        this.name = '';
        this.min = undefined;
        this.max = undefined;
        this.minLength = undefined;
        this.maxLength = undefined;
        this.errorMessageTemplate = undefined;
    }
    TdDynamicInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-input',
                    template: "<div class=\"td-dynamic-input-wrapper\">\n  <mat-form-field class=\"td-dynamic-input-field\">\n    <input #elementInput\n            matInput\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [type]=\"type\"\n            [required]=\"required\"\n            [attr.name]=\"name\"\n            [attr.min]=\"min\"\n            [attr.max]=\"max\"\n            [attr.minLength]=\"minLength\"\n            [attr.maxLength]=\"maxLength\"/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>\n",
                    styles: [".td-dynamic-input-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-input-wrapper .td-dynamic-input-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
                }] }
    ];
    return TdDynamicInputComponent;
}());
export { TdDynamicInputComponent };
if (false) {
    /** @type {?} */
    TdDynamicInputComponent.prototype.control;
    /** @type {?} */
    TdDynamicInputComponent.prototype.label;
    /** @type {?} */
    TdDynamicInputComponent.prototype.hint;
    /** @type {?} */
    TdDynamicInputComponent.prototype.type;
    /** @type {?} */
    TdDynamicInputComponent.prototype.required;
    /** @type {?} */
    TdDynamicInputComponent.prototype.name;
    /** @type {?} */
    TdDynamicInputComponent.prototype.min;
    /** @type {?} */
    TdDynamicInputComponent.prototype.max;
    /** @type {?} */
    TdDynamicInputComponent.prototype.minLength;
    /** @type {?} */
    TdDynamicInputComponent.prototype.maxLength;
    /** @type {?} */
    TdDynamicInputComponent.prototype.errorMessageTemplate;
}
