// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-datepicker/dynamic-datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicDatepickerComponent = /** @class */ (function () {
    function TdDynamicDatepickerComponent() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.type = undefined;
        this.required = undefined;
        this.min = undefined;
        this.max = undefined;
        this.errorMessageTemplate = undefined;
    }
    TdDynamicDatepickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-datepicker',
                    template: "<div class=\"td-dynamic-datepicker-wrapper\">\n  <mat-form-field class=\"td-dynamic-datepicker-field\">\n    <input #elementInput\n            matInput\n            [matDatepicker]=\"dynamicDatePicker\"\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            [name]=\"name\"\n            [min]=\"min\"\n            [max]=\"max\"/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n    <mat-datepicker-toggle matSuffix [for]=\"dynamicDatePicker\"></mat-datepicker-toggle>\n    <mat-datepicker #dynamicDatePicker></mat-datepicker>\n  </mat-form-field>\n</div>\n",
                    styles: [".td-dynamic-datepicker-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-datepicker-wrapper .td-dynamic-datepicker-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
                }] }
    ];
    return TdDynamicDatepickerComponent;
}());
export { TdDynamicDatepickerComponent };
if (false) {
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.control;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.label;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.hint;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.name;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.type;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.required;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.min;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.max;
    /** @type {?} */
    TdDynamicDatepickerComponent.prototype.errorMessageTemplate;
}
