// @covalent/http/esm2015/http.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { NgModule, Injector, InjectionToken } from '@angular/core';
import { HttpClientModule, HttpHandler } from '@angular/common/http';
import { InternalHttpService } from './actions/http.mixin';
import { TdHttpService, TdInterceptorBehaviorService } from './interceptors/http.service';
import { TdURLRegExpInterceptorMatcher } from './interceptors/url-regexp-interceptor-matcher.class';
/** @type {?} */
export const HTTP_CONFIG = new InjectionToken('HTTP_CONFIG');
/**
 * @param {?} handler
 * @param {?} injector
 * @param {?} config
 * @return {?}
 */
export function httpFactory(handler, injector, config) {
    return new TdHttpService(handler, new TdInterceptorBehaviorService(injector, new TdURLRegExpInterceptorMatcher(), config.interceptors));
}
/** @type {?} */
export const HTTP_INTERCEPTOR_PROVIDER = {
    provide: TdHttpService,
    useFactory: httpFactory,
    deps: [HttpHandler, Injector, HTTP_CONFIG],
};
export class CovalentHttpModule {
    /**
     * @param {?} _internalHttpService
     */
    constructor(_internalHttpService) {
        this._internalHttpService = _internalHttpService;
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = { interceptors: [] }) {
        return {
            ngModule: CovalentHttpModule,
            providers: [{
                    provide: HTTP_CONFIG,
                    useValue: config,
                },
                HTTP_INTERCEPTOR_PROVIDER,
            ],
        };
    }
}
CovalentHttpModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                ],
                providers: [
                    InternalHttpService,
                ],
            },] }
];
/** @nocollapse */
CovalentHttpModule.ctorParameters = () => [
    { type: InternalHttpService }
];
if (false) {
    /** @type {?} */
    CovalentHttpModule.prototype._internalHttpService;
}
