// @covalent/http/esm2015/interceptors/http.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
/**
 * @record
 */
export function ITdHttpInterceptorConfig() { }
if (false) {
    /** @type {?} */
    ITdHttpInterceptorConfig.prototype.interceptor;
    /** @type {?} */
    ITdHttpInterceptorConfig.prototype.paths;
}
export class TdInterceptorBehaviorService {
    /**
     * @param {?} _injector
     * @param {?} _httpInterceptorMatcher
     * @param {?} requestInterceptorConfigs
     */
    constructor(_injector, _httpInterceptorMatcher, requestInterceptorConfigs) {
        this._injector = _injector;
        this._httpInterceptorMatcher = _httpInterceptorMatcher;
        this._requestInterceptors = [];
        requestInterceptorConfigs.forEach((config) => {
            this._requestInterceptors.push({
                interceptor: (/** @type {?} */ (_injector.get(config.interceptor))),
                paths: config.paths,
            });
        });
    }
    /**
     * @return {?}
     */
    get requestInterceptors() {
        return this._requestInterceptors;
    }
    /**
     * @return {?}
     */
    get httpInterceptorMatcher() {
        return this._httpInterceptorMatcher;
    }
}
if (false) {
    /** @type {?} */
    TdInterceptorBehaviorService.prototype._requestInterceptors;
    /** @type {?} */
    TdInterceptorBehaviorService.prototype._injector;
    /** @type {?} */
    TdInterceptorBehaviorService.prototype._httpInterceptorMatcher;
}
export class TdHttpService extends HttpClient {
    /**
     * @param {?} _handler
     * @param {?} _interceptorBehavior
     */
    constructor(_handler, _interceptorBehavior) {
        super(_handler);
        this._handler = _handler;
        this._interceptorBehavior = _interceptorBehavior;
    }
    /**
     * @param {?} first
     * @param {?=} url
     * @param {?=} options
     * @return {?}
     */
    request(first, url, options = {}) {
        /** @type {?} */
        let interceptors = this._interceptorBehavior.requestInterceptors.filter((mapping) => {
            return this._interceptorBehavior.httpInterceptorMatcher.matches({ url: url }, mapping);
        }).map((mapping) => {
            return mapping.interceptor;
        });
        return this._setupRequest(first, url, options, interceptors);
    }
    /**
     * @param {?} method
     * @param {?} url
     * @param {?} options
     * @param {?} interceptors
     * @return {?}
     */
    _setupRequest(method, url, options, interceptors) {
        try {
            options = this._requestResolve(options, interceptors);
        }
        catch (e) {
            return new Observable((subscriber) => {
                subscriber.error(e);
            });
        }
        return this._handleResponseResolve(super.request((/** @type {?} */ (method)), url, options), interceptors);
    }
    /**
     * @param {?} options
     * @param {?} interceptors
     * @return {?}
     */
    _requestResolve(options, interceptors) {
        interceptors.forEach((interceptor) => {
            if (interceptor.handleOptions) {
                options = interceptor.handleOptions(options);
            }
        });
        return options;
    }
    /**
     * @param {?} obs
     * @param {?} interceptors
     * @return {?}
     */
    _handleResponseResolve(obs, interceptors) {
        interceptors.reverse().forEach((interceptor) => {
            if (interceptor.handleResponse) {
                obs = interceptor.handleResponse(obs);
            }
        });
        return obs;
    }
}
if (false) {
    /** @type {?} */
    TdHttpService.prototype._handler;
    /** @type {?} */
    TdHttpService.prototype._interceptorBehavior;
}
