// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-select/dynamic-select.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicSelectComponent = /** @class */ (function () {
    function TdDynamicSelectComponent() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.required = undefined;
        this.selections = undefined;
        this.multiple = undefined;
        this.errorMessageTemplate = undefined;
    }
    TdDynamicSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-select',
                    template: "<div class=\"td-dynamic-select-wrapper\">\n  <mat-form-field class=\"td-dynamic-select-field\">\n    <mat-select [formControl]=\"control\"\n                [placeholder]=\"label\"\n                [required]=\"required\"\n                [attr.name]=\"name\"\n                [multiple]=\"multiple\">\n      <mat-option *ngFor=\"let selection of selections\" [value]=\"selection.value || selection\">{{selection.label || selection}}</mat-option>\n    </mat-select>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>\n",
                    styles: [".td-dynamic-select-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-select-wrapper .td-dynamic-select-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
                }] }
    ];
    return TdDynamicSelectComponent;
}());
export { TdDynamicSelectComponent };
if (false) {
    /** @type {?} */
    TdDynamicSelectComponent.prototype.control;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.label;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.hint;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.name;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.required;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.selections;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.multiple;
    /** @type {?} */
    TdDynamicSelectComponent.prototype.errorMessageTemplate;
}
