// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-textarea/dynamic-textarea.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicTextareaComponent = /** @class */ (function () {
    function TdDynamicTextareaComponent() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.required = undefined;
        this.errorMessageTemplate = undefined;
    }
    TdDynamicTextareaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-textarea',
                    template: "<div class=\"td-dynamic-textarea-wrapper\">\n  <mat-form-field class=\"td-dynamic-textarea-field\">\n    <textarea #elementInput\n            matInput\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            [attr.name]=\"name\"\n            rows=\"4\">\n    </textarea>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>",
                    styles: [".td-dynamic-textarea-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-textarea-wrapper .td-dynamic-textarea-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
                }] }
    ];
    return TdDynamicTextareaComponent;
}());
export { TdDynamicTextareaComponent };
if (false) {
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.control;
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.label;
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.hint;
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.name;
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.required;
    /** @type {?} */
    TdDynamicTextareaComponent.prototype.errorMessageTemplate;
}
