// @covalent/dynamic-forms/esm2015/dynamic-forms.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectionStrategy, ChangeDetectorRef, ContentChildren, QueryList } from '@angular/core';
import { FormBuilder } from '@angular/forms';
import { TdDynamicFormsService } from './services/dynamic-forms.service';
import { TdDynamicFormsErrorTemplate } from './dynamic-element.component';
import { timer, Subject } from 'rxjs';
import { takeUntil, filter } from 'rxjs/operators';
export class TdDynamicFormsComponent {
    /**
     * @param {?} _formBuilder
     * @param {?} _dynamicFormsService
     * @param {?} _changeDetectorRef
     */
    constructor(_formBuilder, _dynamicFormsService, _changeDetectorRef) {
        this._formBuilder = _formBuilder;
        this._dynamicFormsService = _dynamicFormsService;
        this._changeDetectorRef = _changeDetectorRef;
        this._renderedElements = [];
        this._templateMap = new Map();
        this._destroy$ = new Subject();
        this._destroyControl$ = new Subject();
        this.dynamicForm = this._formBuilder.group({});
    }
    /**
     * elements: ITdDynamicElementConfig[]
     * JS Object that will render the elements depending on its config.
     * [name] property is required.
     * @param {?} elements
     * @return {?}
     */
    set elements(elements) {
        if (elements) {
            this._elements = elements;
        }
        else {
            this._elements = [];
        }
        this._rerenderElements();
    }
    /**
     * @return {?}
     */
    get elements() {
        return this._renderedElements;
    }
    /**
     * Getter property for dynamic [FormGroup].
     * @return {?}
     */
    get form() {
        return this.dynamicForm;
    }
    /**
     * Getter property for [valid] of dynamic [FormGroup].
     * @return {?}
     */
    get valid() {
        if (this.dynamicForm) {
            return this.dynamicForm.valid;
        }
        return false;
    }
    /**
     * Getter property for [value] of dynamic [FormGroup].
     * @return {?}
     */
    get value() {
        if (this.dynamicForm) {
            return this.dynamicForm.value;
        }
        return {};
    }
    /**
     * Getter property for [errors] of dynamic [FormGroup].
     * @return {?}
     */
    get errors() {
        if (this.dynamicForm) {
            /** @type {?} */
            let errors = {};
            for (let name in this.dynamicForm.controls) {
                errors[name] = this.dynamicForm.controls[name].errors;
            }
            return errors;
        }
        return {};
    }
    /**
     * Getter property for [controls] of dynamic [FormGroup].
     * @return {?}
     */
    get controls() {
        if (this.dynamicForm) {
            return this.dynamicForm.controls;
        }
        return {};
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._updateErrorTemplates();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
        this._destroyControl$.complete();
    }
    /**
     * Refreshes the form and rerenders all validator/element modifications.
     * @return {?}
     */
    refresh() {
        this._rerenderElements();
        this._updateErrorTemplates();
    }
    /**
     * Getter method for error template references
     * @param {?} name
     * @return {?}
     */
    getErrorTemplateRef(name) {
        return this._templateMap.get(name);
    }
    /**
     * Loads error templates and sets them in a map for faster access.
     * @return {?}
     */
    _updateErrorTemplates() {
        this._templateMap = new Map();
        for (let i = 0; i < this._errorTemplates.toArray().length; i++) {
            this._templateMap.set(this._errorTemplates.toArray()[i].tdDynamicFormsError, this._errorTemplates.toArray()[i].templateRef);
        }
    }
    /**
     * @return {?}
     */
    _rerenderElements() {
        this._clearRemovedElements();
        this._renderedElements = [];
        /** @type {?} */
        let duplicates = [];
        this._elements.forEach((elem) => {
            this._dynamicFormsService.validateDynamicElementName(elem.name);
            if (duplicates.indexOf(elem.name) > -1) {
                throw new Error(`Dynamic element name: "${elem.name}" is duplicated`);
            }
            duplicates.push(elem.name);
            /** @type {?} */
            let dynamicElement = this.dynamicForm.get(elem.name);
            if (!dynamicElement) {
                this.dynamicForm.addControl(elem.name, this._dynamicFormsService.createFormControl(elem));
                this._subscribeToControlStatusChanges(elem.name);
            }
            else {
                dynamicElement.setValue(elem.default);
                dynamicElement.markAsPristine();
                dynamicElement.markAsUntouched();
                if (elem.disabled) {
                    dynamicElement.disable();
                }
                else {
                    dynamicElement.enable();
                }
                dynamicElement.setValidators(this._dynamicFormsService.createValidators(elem));
            }
            // copy objects so they are only changes when calling this method
            this._renderedElements.push(Object.assign({}, elem));
        });
        // call a change detection since the whole form might change
        this._changeDetectorRef.detectChanges();
        timer().toPromise().then(() => {
            // call a markForCheck so elements are rendered correctly in OnPush
            this._changeDetectorRef.markForCheck();
        });
    }
    /**
     * @return {?}
     */
    _clearRemovedElements() {
        for (let i = 0; i < this._renderedElements.length; i++) {
            for (let j = 0; j < this._elements.length; j++) {
                // check if the name of the element is still there removed
                if (this._renderedElements[i].name === this._elements[j].name) {
                    delete this._renderedElements[i];
                    break;
                }
            }
        }
        // remove elements that were removed from the array
        this._renderedElements.forEach((elem) => {
            this._destroyControl$.next(elem.name);
            this.dynamicForm.removeControl(elem.name);
        });
    }
    // Updates component when manually adding errors to controls
    /**
     * @param {?} elementName
     * @return {?}
     */
    _subscribeToControlStatusChanges(elementName) {
        /** @type {?} */
        const control = this.controls[elementName];
        /** @type {?} */
        const controlDestroyed$ = this._destroyControl$
            .pipe(filter((destroyedElementName) => destroyedElementName === elementName));
        control.statusChanges
            .pipe(takeUntil(this._destroy$), takeUntil(controlDestroyed$)).subscribe(() => {
            this._changeDetectorRef.markForCheck();
        });
    }
}
TdDynamicFormsComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-forms',
                template: "<form [formGroup]=\"dynamicForm\" novalidate>\n  <div class=\"td-dynamic-form-wrapper\">\n    <ng-template let-element ngFor [ngForOf]=\"elements\">\n      <div class=\"td-dynamic-element-wrapper\"\n          [style.max-width.%]=\"element.flex ? element.flex : 100\"\n          [style.flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-ms-flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-webkit-box-flex]=\"1\">\n        <td-dynamic-element\n          #dynamicElement\n          *ngIf=\"dynamicForm.controls[element.name]\"\n          [formControlName]=\"element.name\"\n          [dynamicControl]=\"dynamicForm.controls[element.name]\"\n          [id]=\"element.name\"\n          [name]=\"element.name\"\n          [label]=\"element.label || element.name\"\n          [hint]=\"element.hint\"\n          [type]=\"element.type\"\n          [required]=\"element.required\"\n          [min]=\"element.min\"\n          [max]=\"element.max\"\n          [minLength]=\"element.minLength\"\n          [maxLength]=\"element.maxLength\"\n          [selections]=\"element.selections\"\n          [multiple]=\"element.multiple\"\n          [errorMessageTemplate]=\"getErrorTemplateRef(element.name)\">\n        </td-dynamic-element>\n      </div>\n    </ng-template>\n  </div>\n  <ng-content></ng-content>\n</form>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".td-dynamic-form-wrapper{-ms-flex-wrap:wrap;flex-wrap:wrap;box-sizing:border-box;display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:center;align-items:center;-ms-flex-line-pack:center;align-content:center;max-width:100%;-ms-flex-pack:start;justify-content:start}.td-dynamic-form-wrapper ::ng-deep .mat-form-field-infix{width:auto}.td-dynamic-form-wrapper ::ng-deep .td-dynamic-element-hint{font-size:75%;display:block}.td-dynamic-form-wrapper .td-dynamic-element-wrapper{max-height:100%;box-sizing:border-box;position:relative;padding:4px 4px 8px}"]
            }] }
];
/** @nocollapse */
TdDynamicFormsComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: TdDynamicFormsService },
    { type: ChangeDetectorRef }
];
TdDynamicFormsComponent.propDecorators = {
    _errorTemplates: [{ type: ContentChildren, args: [TdDynamicFormsErrorTemplate,] }],
    elements: [{ type: Input, args: ['elements',] }]
};
if (false) {
    /** @type {?} */
    TdDynamicFormsComponent.prototype._renderedElements;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._elements;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._templateMap;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._destroy$;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._destroyControl$;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._errorTemplates;
    /** @type {?} */
    TdDynamicFormsComponent.prototype.dynamicForm;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._formBuilder;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._dynamicFormsService;
    /** @type {?} */
    TdDynamicFormsComponent.prototype._changeDetectorRef;
}
