// @covalent/dynamic-forms/fesm2015/covalent-dynamic-forms.js
import { CommonModule } from '@angular/common';
import { MatInputModule } from '@angular/material/input';
import { MatSelectModule } from '@angular/material/select';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatSliderModule } from '@angular/material/slider';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { CovalentFileModule } from '@covalent/core/file';
import { Component, ChangeDetectorRef, Injectable, SkipSelf, Optional, Directive, Input, HostBinding, TemplateRef, Type, ViewChild, ViewContainerRef, ComponentFactoryResolver, forwardRef, ChangeDetectionStrategy, ContentChildren, NgModule } from '@angular/core';
import { Validators, FormControl, NG_VALUE_ACCESSOR, FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { TemplatePortalDirective } from '@angular/cdk/portal';
import { mixinControlValueAccessor, CovalentCommonModule } from '@covalent/core/common';
import { timer, Subject } from 'rxjs';
import { takeUntil, filter } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicInputComponent {
    constructor() {
        this.label = '';
        this.hint = '';
        this.type = undefined;
        this.required = undefined;
        this.name = '';
        this.min = undefined;
        this.max = undefined;
        this.minLength = undefined;
        this.maxLength = undefined;
        this.errorMessageTemplate = undefined;
    }
}
TdDynamicInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-input',
                template: "<div class=\"td-dynamic-input-wrapper\">\n  <mat-form-field class=\"td-dynamic-input-field\">\n    <input #elementInput\n            matInput\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [type]=\"type\"\n            [required]=\"required\"\n            [attr.name]=\"name\"\n            [attr.min]=\"min\"\n            [attr.max]=\"max\"\n            [attr.minLength]=\"minLength\"\n            [attr.maxLength]=\"maxLength\"/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>\n",
                styles: [".td-dynamic-input-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-input-wrapper .td-dynamic-input-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicFileInputComponent {
    constructor() {
        this.required = undefined;
        this.label = '';
        this.name = '';
        this.hint = '';
        this.errorMessageTemplate = undefined;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    _handlefileDrop(value) {
        this.control.setValue(value);
    }
}
TdDynamicFileInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-file-input',
                template: "<div class=\"td-dynamic-file-input-wrapper\">\n  <mat-form-field tdFileDrop\n                  class=\"td-dynamic-file-input-field\"\n                  floatLabel=\"never\"\n                  [disabled]=\"control?.disabled\"\n                  (fileDrop)=\"_handlefileDrop($event)\"\n                  (click)=\"!control?.disabled && fileInput.inputElement.click()\"\n                  (keyup.enter)=\"!control?.disabled && fileInput.inputElement.click()\"\n                  (keyup.delete)=\"fileInput.clear()\"\n                  (keyup.backspace)=\"fileInput.clear()\">\n    <input matInput\n          [value]=\"control?.value?.name\"\n          [placeholder]=\"label\"\n          [attr.name]=\"name\"\n          [disabled]=\"control?.disabled\"\n          readonly/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n  <button mat-icon-button *ngIf=\"control.value\" (click)=\"fileInput.clear()\" (keyup.enter)=\"fileInput.clear()\">\n    <mat-icon>cancel</mat-icon>\n  </button>\n  <td-file-input class=\"td-file-input\"\n                  #fileInput\n                  [formControl]=\"control\">\n    <mat-icon>folder</mat-icon>\n    <span>{{ label }}</span>\n  </td-file-input>\n</div>",
                styles: [".td-dynamic-file-input-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-file-input-wrapper .td-dynamic-file-input-field{-ms-flex:1;flex:1;box-sizing:border-box}.td-file-input{margin-left:10px}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicTextareaComponent {
    constructor() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.required = undefined;
        this.errorMessageTemplate = undefined;
    }
}
TdDynamicTextareaComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-textarea',
                template: "<div class=\"td-dynamic-textarea-wrapper\">\n  <mat-form-field class=\"td-dynamic-textarea-field\">\n    <textarea #elementInput\n            matInput\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            [attr.name]=\"name\"\n            rows=\"4\">\n    </textarea>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>",
                styles: [".td-dynamic-textarea-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-textarea-wrapper .td-dynamic-textarea-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicSlideToggleComponent {
    constructor() {
        this.label = '';
        this.name = '';
        this.hint = '';
        this.required = false;
    }
}
TdDynamicSlideToggleComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-slide-toggle',
                template: "<div class=\"td-dynamic-slide-toggle-wrapper\">\n  <mat-slide-toggle [formControl]=\"control\"\n                    [attr.name]=\"name\"\n                    [required]=\"required\">\n    {{label}}\n  </mat-slide-toggle>\n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                styles: [""]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicCheckboxComponent {
    constructor() {
        this.label = '';
        this.name = '';
        this.hint = '';
        this.required = false;
    }
}
TdDynamicCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-checkbox',
                template: "<div class=\"td-dynamic-checkbox-wrapper\">\n  <mat-checkbox [formControl]=\"control\"\n                [name]=\"name\"\n                [required]=\"required\">\n    {{label}}\n  </mat-checkbox>\n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                styles: [""]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicSliderComponent {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this.label = '';
        this.required = undefined;
        this.name = '';
        this.hint = '';
        this.min = undefined;
        this.max = undefined;
    }
    /**
     * @return {?}
     */
    _handleBlur() {
        setTimeout(() => {
            this._changeDetectorRef.markForCheck();
        });
    }
}
TdDynamicSliderComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-slider',
                template: "<div class=\"td-dynamic-slider-wrapper mat-form-field mat-form-field-can-float mat-form-field-should-float\"\n      [class.mat-focused]=\"slider._isActive\">\n  <span class=\"mat-form-field-label-wrapper\">\n    <label class=\"mat-form-field-label mat-primary td-slider-label\">\n      {{label}}\n      <span *ngIf=\"required && !control?.disabled\" class=\"mat-form-field-required-marker\">*</span>\n    </label>\n  </span>\n  <div class=\"td-dynamic-slider-field\">\n    <mat-slider #slider\n                class=\"td-dynamic-slider\"\n                [formControl]=\"control\"\n                [attr.name]=\"name\"\n                [min]=\"min\"\n                [max]=\"max\"\n                thumbLabel\n                tickInterval=\"auto\"\n                [required]=\"required\"\n                (blur)=\"_handleBlur()\">\n    </mat-slider>\n  </div>  \n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                styles: [":host .td-dynamic-slider-wrapper{display:block}.td-dynamic-slider-field{position:relative;-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-slider-field .td-dynamic-slider{-ms-flex:1;flex:1}"]
            }] }
];
/** @nocollapse */
TdDynamicSliderComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicSelectComponent {
    constructor() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.required = undefined;
        this.selections = undefined;
        this.multiple = undefined;
        this.errorMessageTemplate = undefined;
    }
}
TdDynamicSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-select',
                template: "<div class=\"td-dynamic-select-wrapper\">\n  <mat-form-field class=\"td-dynamic-select-field\">\n    <mat-select [formControl]=\"control\"\n                [placeholder]=\"label\"\n                [required]=\"required\"\n                [attr.name]=\"name\"\n                [multiple]=\"multiple\">\n      <mat-option *ngFor=\"let selection of selections\" [value]=\"selection.value || selection\">{{selection.label || selection}}</mat-option>\n    </mat-select>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n  </mat-form-field>\n</div>\n",
                styles: [".td-dynamic-select-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-select-wrapper .td-dynamic-select-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicDatepickerComponent {
    constructor() {
        this.label = '';
        this.hint = '';
        this.name = '';
        this.type = undefined;
        this.required = undefined;
        this.min = undefined;
        this.max = undefined;
        this.errorMessageTemplate = undefined;
    }
}
TdDynamicDatepickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-datepicker',
                template: "<div class=\"td-dynamic-datepicker-wrapper\">\n  <mat-form-field class=\"td-dynamic-datepicker-field\">\n    <input #elementInput\n            matInput\n            [matDatepicker]=\"dynamicDatePicker\"\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            [name]=\"name\"\n            [min]=\"min\"\n            [max]=\"max\"/>\n    <mat-hint>{{hint}}</mat-hint>\n    <mat-error>\n      <ng-template\n        [ngTemplateOutlet]=\"errorMessageTemplate\"\n        [ngTemplateOutletContext]=\"{control: control, errors: control?.errors}\">\n      </ng-template>\n    </mat-error>\n    <mat-datepicker-toggle matSuffix [for]=\"dynamicDatePicker\"></mat-datepicker-toggle>\n    <mat-datepicker #dynamicDatePicker></mat-datepicker>\n  </mat-form-field>\n</div>\n",
                styles: [".td-dynamic-datepicker-wrapper{-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-datepicker-wrapper .td-dynamic-datepicker-field{-ms-flex:1;flex:1;box-sizing:border-box}"]
            }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/** @enum {string} */
const TdDynamicType = {
    Text: 'text',
    Boolean: 'boolean',
    Number: 'number',
    Array: 'array',
    Date: 'date',
};
/** @enum {string} */
const TdDynamicElement = {
    Input: 'input',
    Datepicker: 'datepicker',
    Password: 'password',
    Textarea: 'textarea',
    Slider: 'slider',
    SlideToggle: 'slide-toggle',
    Checkbox: 'checkbox',
    Select: 'select',
    FileInput: 'file-input',
};
/** @type {?} */
const DYNAMIC_ELEMENT_NAME_REGEX = /^[^0-9][^\@]*$/;
class TdDynamicFormsService {
    /**
     * Method to validate if the [name] is a proper element name.
     * Throws error if name is not valid.
     * @param {?} name
     * @return {?}
     */
    validateDynamicElementName(name) {
        if (!DYNAMIC_ELEMENT_NAME_REGEX.test(name)) {
            throw new Error('Dynamic element name: "${name}" is not valid.');
        }
    }
    /**
     * Gets component to be rendered depending on [TdDynamicElement | TdDynamicType]
     * Throws error if it does not exists or not supported.
     * @param {?} element
     * @return {?}
     */
    getDynamicElement(element) {
        switch (element) {
            case TdDynamicType.Text:
            case TdDynamicType.Number:
            case TdDynamicElement.Input:
            case TdDynamicElement.Password:
                return TdDynamicInputComponent;
            case TdDynamicElement.Textarea:
                return TdDynamicTextareaComponent;
            case TdDynamicType.Boolean:
            case TdDynamicElement.SlideToggle:
                return TdDynamicSlideToggleComponent;
            case TdDynamicElement.Checkbox:
                return TdDynamicCheckboxComponent;
            case TdDynamicElement.Slider:
                return TdDynamicSliderComponent;
            case TdDynamicType.Array:
            case TdDynamicElement.Select:
                return TdDynamicSelectComponent;
            case TdDynamicElement.FileInput:
                return TdDynamicFileInputComponent;
            case TdDynamicElement.Datepicker:
            case TdDynamicType.Date:
                return TdDynamicDatepickerComponent;
            default:
                throw new Error(`Error: type ${element} does not exist or not supported.`);
        }
    }
    /**
     * Creates form control for element depending [ITdDynamicElementConfig] properties.
     * @param {?} config
     * @return {?}
     */
    createFormControl(config) {
        /** @type {?} */
        let validator = this.createValidators(config);
        return new FormControl({ value: config.default, disabled: config.disabled }, validator);
    }
    /**
     * Creates form validationdepending [ITdDynamicElementConfig] properties.
     * @param {?} config
     * @return {?}
     */
    createValidators(config) {
        /** @type {?} */
        let validator;
        if (config.required) {
            validator = Validators.required;
        }
        if (config.max || config.max === 0) {
            validator = Validators.compose([validator, Validators.max(parseFloat(config.max))]);
        }
        if (config.min || config.min === 0) {
            validator = Validators.compose([validator, Validators.min(parseFloat(config.min))]);
        }
        if (config.maxLength || config.maxLength === 0) {
            validator = Validators.compose([validator, Validators.maxLength(parseFloat(config.maxLength))]);
        }
        if (config.minLength || config.minLength === 0) {
            validator = Validators.compose([validator, Validators.minLength(parseFloat(config.minLength))]);
        }
        // Add provided custom validators to the validator function
        if (config.validators) {
            config.validators.forEach((validatorConfig) => {
                validator = Validators.compose([validator, validatorConfig.validator]);
            });
        }
        return validator;
    }
}
TdDynamicFormsService.decorators = [
    { type: Injectable }
];
/**
 * @param {?} parent
 * @return {?}
 */
function DYNAMIC_FORMS_PROVIDER_FACTORY(parent) {
    return parent || new TdDynamicFormsService();
}
/** @type {?} */
const DYNAMIC_FORMS_PROVIDER = {
    // If there is already a service available, use that. Otherwise, provide a new one.
    provide: TdDynamicFormsService,
    deps: [[new Optional(), new SkipSelf(), TdDynamicFormsService]],
    useFactory: DYNAMIC_FORMS_PROVIDER_FACTORY,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicElementBase {
    /**
     * @param {?} _changeDetectorRef
     */
    constructor(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
    }
}
/* tslint:disable-next-line */
/** @type {?} */
const _TdDynamicElementMixinBase = mixinControlValueAccessor(TdDynamicElementBase);
class TdDynamicFormsErrorTemplate extends TemplatePortalDirective {
    /**
     * @param {?} templateRef
     * @param {?} viewContainerRef
     */
    constructor(templateRef, viewContainerRef) {
        super(templateRef, viewContainerRef);
    }
}
TdDynamicFormsErrorTemplate.decorators = [
    { type: Directive, args: [{ selector: '[tdDynamicFormsError]ng-template' },] }
];
/** @nocollapse */
TdDynamicFormsErrorTemplate.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef }
];
TdDynamicFormsErrorTemplate.propDecorators = {
    tdDynamicFormsError: [{ type: Input }]
};
class TdDynamicElementDirective {
    /**
     * @param {?} viewContainer
     */
    constructor(viewContainer) {
        this.viewContainer = viewContainer;
    }
}
TdDynamicElementDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tdDynamicContainer]',
            },] }
];
/** @nocollapse */
TdDynamicElementDirective.ctorParameters = () => [
    { type: ViewContainerRef }
];
class TdDynamicElementComponent extends _TdDynamicElementMixinBase {
    /**
     * @param {?} _componentFactoryResolver
     * @param {?} _dynamicFormsService
     * @param {?} _changeDetectorRef
     */
    constructor(_componentFactoryResolver, _dynamicFormsService, _changeDetectorRef) {
        super(_changeDetectorRef);
        this._componentFactoryResolver = _componentFactoryResolver;
        this._dynamicFormsService = _dynamicFormsService;
        /**
         * Sets label to be displayed.
         */
        this.label = '';
        /**
         * Sets hint to be displayed.
         */
        this.hint = '';
        /**
         * Sets name to be displayed as attribute.
         */
        this.name = '';
        /**
         * Sets type or element of element to be rendered.
         * Throws error if does not exist or no supported.
         */
        this.type = undefined;
        /**
         * Sets required validation checkup (if supported by element).
         */
        this.required = undefined;
        /**
         * Sets min validation checkup (if supported by element).
         */
        this.min = undefined;
        /**
         * Sets max validation checkup (if supported by element).
         */
        this.max = undefined;
        /**
         * Sets minLength validation checkup (if supported by element).
         */
        this.minLength = undefined;
        /**
         * Sets maxLength validation checkup (if supported by element).
         */
        this.maxLength = undefined;
        /**
         * Sets selections for array elements (if supported by element).
         */
        this.selections = undefined;
        /**
         * Sets multiple property for array elements (if supported by element).
         */
        this.multiple = undefined;
        /**
         * Sets error message template so it can be injected into dynamic components.
         */
        this.errorMessageTemplate = undefined;
    }
    /**
     * @return {?}
     */
    get maxAttr() {
        return this.max;
    }
    /**
     * @return {?}
     */
    get minAttr() {
        return this.min;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        let component = (/** @type {?} */ (this.type)) instanceof Type ? this.type : this._dynamicFormsService.getDynamicElement(this.type);
        /** @type {?} */
        let ref = this._componentFactoryResolver.
            resolveComponentFactory(component)
            .create(this.childElement.viewContainer.injector);
        this.childElement.viewContainer.insert(ref.hostView);
        this._instance = ref.instance;
        this._instance.control = this.dynamicControl;
        this._instance.label = this.label;
        this._instance.hint = this.hint;
        this._instance.name = this.name;
        this._instance.type = this.type;
        this._instance.value = this.value;
        this._instance.required = this.required;
        this._instance.min = this.min;
        this._instance.max = this.max;
        this._instance.minLength = this.minLength;
        this._instance.maxLength = this.maxLength;
        this._instance.selections = this.selections;
        this._instance.multiple = this.multiple;
        this._instance.errorMessageTemplate = this.errorMessageTemplate;
    }
    /**
     * Reassign any inputs that have changed
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this._instance) {
            for (let prop in changes) {
                this._instance[prop] = changes[prop].currentValue;
            }
        }
    }
}
TdDynamicElementComponent.decorators = [
    { type: Component, args: [{
                providers: [TdDynamicFormsService, {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => TdDynamicElementComponent),
                        multi: true,
                    }],
                selector: 'td-dynamic-element',
                template: '<div tdDynamicContainer></div>'
            }] }
];
/** @nocollapse */
TdDynamicElementComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: TdDynamicFormsService },
    { type: ChangeDetectorRef }
];
TdDynamicElementComponent.propDecorators = {
    dynamicControl: [{ type: Input }],
    label: [{ type: Input }],
    hint: [{ type: Input }],
    name: [{ type: Input }],
    type: [{ type: Input }],
    required: [{ type: Input }],
    min: [{ type: Input }],
    max: [{ type: Input }],
    minLength: [{ type: Input }],
    maxLength: [{ type: Input }],
    selections: [{ type: Input }],
    multiple: [{ type: Input }],
    errorMessageTemplate: [{ type: Input }],
    childElement: [{ type: ViewChild, args: [TdDynamicElementDirective,] }],
    maxAttr: [{ type: HostBinding, args: ['attr.max',] }],
    minAttr: [{ type: HostBinding, args: ['attr.min',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
class TdDynamicFormsComponent {
    /**
     * @param {?} _formBuilder
     * @param {?} _dynamicFormsService
     * @param {?} _changeDetectorRef
     */
    constructor(_formBuilder, _dynamicFormsService, _changeDetectorRef) {
        this._formBuilder = _formBuilder;
        this._dynamicFormsService = _dynamicFormsService;
        this._changeDetectorRef = _changeDetectorRef;
        this._renderedElements = [];
        this._templateMap = new Map();
        this._destroy$ = new Subject();
        this._destroyControl$ = new Subject();
        this.dynamicForm = this._formBuilder.group({});
    }
    /**
     * elements: ITdDynamicElementConfig[]
     * JS Object that will render the elements depending on its config.
     * [name] property is required.
     * @param {?} elements
     * @return {?}
     */
    set elements(elements) {
        if (elements) {
            this._elements = elements;
        }
        else {
            this._elements = [];
        }
        this._rerenderElements();
    }
    /**
     * @return {?}
     */
    get elements() {
        return this._renderedElements;
    }
    /**
     * Getter property for dynamic [FormGroup].
     * @return {?}
     */
    get form() {
        return this.dynamicForm;
    }
    /**
     * Getter property for [valid] of dynamic [FormGroup].
     * @return {?}
     */
    get valid() {
        if (this.dynamicForm) {
            return this.dynamicForm.valid;
        }
        return false;
    }
    /**
     * Getter property for [value] of dynamic [FormGroup].
     * @return {?}
     */
    get value() {
        if (this.dynamicForm) {
            return this.dynamicForm.value;
        }
        return {};
    }
    /**
     * Getter property for [errors] of dynamic [FormGroup].
     * @return {?}
     */
    get errors() {
        if (this.dynamicForm) {
            /** @type {?} */
            let errors = {};
            for (let name in this.dynamicForm.controls) {
                errors[name] = this.dynamicForm.controls[name].errors;
            }
            return errors;
        }
        return {};
    }
    /**
     * Getter property for [controls] of dynamic [FormGroup].
     * @return {?}
     */
    get controls() {
        if (this.dynamicForm) {
            return this.dynamicForm.controls;
        }
        return {};
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._updateErrorTemplates();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
        this._destroyControl$.complete();
    }
    /**
     * Refreshes the form and rerenders all validator/element modifications.
     * @return {?}
     */
    refresh() {
        this._rerenderElements();
        this._updateErrorTemplates();
    }
    /**
     * Getter method for error template references
     * @param {?} name
     * @return {?}
     */
    getErrorTemplateRef(name) {
        return this._templateMap.get(name);
    }
    /**
     * Loads error templates and sets them in a map for faster access.
     * @return {?}
     */
    _updateErrorTemplates() {
        this._templateMap = new Map();
        for (let i = 0; i < this._errorTemplates.toArray().length; i++) {
            this._templateMap.set(this._errorTemplates.toArray()[i].tdDynamicFormsError, this._errorTemplates.toArray()[i].templateRef);
        }
    }
    /**
     * @return {?}
     */
    _rerenderElements() {
        this._clearRemovedElements();
        this._renderedElements = [];
        /** @type {?} */
        let duplicates = [];
        this._elements.forEach((elem) => {
            this._dynamicFormsService.validateDynamicElementName(elem.name);
            if (duplicates.indexOf(elem.name) > -1) {
                throw new Error(`Dynamic element name: "${elem.name}" is duplicated`);
            }
            duplicates.push(elem.name);
            /** @type {?} */
            let dynamicElement = this.dynamicForm.get(elem.name);
            if (!dynamicElement) {
                this.dynamicForm.addControl(elem.name, this._dynamicFormsService.createFormControl(elem));
                this._subscribeToControlStatusChanges(elem.name);
            }
            else {
                dynamicElement.setValue(elem.default);
                dynamicElement.markAsPristine();
                dynamicElement.markAsUntouched();
                if (elem.disabled) {
                    dynamicElement.disable();
                }
                else {
                    dynamicElement.enable();
                }
                dynamicElement.setValidators(this._dynamicFormsService.createValidators(elem));
            }
            // copy objects so they are only changes when calling this method
            this._renderedElements.push(Object.assign({}, elem));
        });
        // call a change detection since the whole form might change
        this._changeDetectorRef.detectChanges();
        timer().toPromise().then(() => {
            // call a markForCheck so elements are rendered correctly in OnPush
            this._changeDetectorRef.markForCheck();
        });
    }
    /**
     * @return {?}
     */
    _clearRemovedElements() {
        for (let i = 0; i < this._renderedElements.length; i++) {
            for (let j = 0; j < this._elements.length; j++) {
                // check if the name of the element is still there removed
                if (this._renderedElements[i].name === this._elements[j].name) {
                    delete this._renderedElements[i];
                    break;
                }
            }
        }
        // remove elements that were removed from the array
        this._renderedElements.forEach((elem) => {
            this._destroyControl$.next(elem.name);
            this.dynamicForm.removeControl(elem.name);
        });
    }
    // Updates component when manually adding errors to controls
    /**
     * @param {?} elementName
     * @return {?}
     */
    _subscribeToControlStatusChanges(elementName) {
        /** @type {?} */
        const control = this.controls[elementName];
        /** @type {?} */
        const controlDestroyed$ = this._destroyControl$
            .pipe(filter((destroyedElementName) => destroyedElementName === elementName));
        control.statusChanges
            .pipe(takeUntil(this._destroy$), takeUntil(controlDestroyed$)).subscribe(() => {
            this._changeDetectorRef.markForCheck();
        });
    }
}
TdDynamicFormsComponent.decorators = [
    { type: Component, args: [{
                selector: 'td-dynamic-forms',
                template: "<form [formGroup]=\"dynamicForm\" novalidate>\n  <div class=\"td-dynamic-form-wrapper\">\n    <ng-template let-element ngFor [ngForOf]=\"elements\">\n      <div class=\"td-dynamic-element-wrapper\"\n          [style.max-width.%]=\"element.flex ? element.flex : 100\"\n          [style.flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-ms-flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-webkit-box-flex]=\"1\">\n        <td-dynamic-element\n          #dynamicElement\n          *ngIf=\"dynamicForm.controls[element.name]\"\n          [formControlName]=\"element.name\"\n          [dynamicControl]=\"dynamicForm.controls[element.name]\"\n          [id]=\"element.name\"\n          [name]=\"element.name\"\n          [label]=\"element.label || element.name\"\n          [hint]=\"element.hint\"\n          [type]=\"element.type\"\n          [required]=\"element.required\"\n          [min]=\"element.min\"\n          [max]=\"element.max\"\n          [minLength]=\"element.minLength\"\n          [maxLength]=\"element.maxLength\"\n          [selections]=\"element.selections\"\n          [multiple]=\"element.multiple\"\n          [errorMessageTemplate]=\"getErrorTemplateRef(element.name)\">\n        </td-dynamic-element>\n      </div>\n    </ng-template>\n  </div>\n  <ng-content></ng-content>\n</form>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".td-dynamic-form-wrapper{-ms-flex-wrap:wrap;flex-wrap:wrap;box-sizing:border-box;display:-ms-flexbox;display:flex;-ms-flex-direction:row;flex-direction:row;-ms-flex-align:center;align-items:center;-ms-flex-line-pack:center;align-content:center;max-width:100%;-ms-flex-pack:start;justify-content:start}.td-dynamic-form-wrapper ::ng-deep .mat-form-field-infix{width:auto}.td-dynamic-form-wrapper ::ng-deep .td-dynamic-element-hint{font-size:75%;display:block}.td-dynamic-form-wrapper .td-dynamic-element-wrapper{max-height:100%;box-sizing:border-box;position:relative;padding:4px 4px 8px}"]
            }] }
];
/** @nocollapse */
TdDynamicFormsComponent.ctorParameters = () => [
    { type: FormBuilder },
    { type: TdDynamicFormsService },
    { type: ChangeDetectorRef }
];
TdDynamicFormsComponent.propDecorators = {
    _errorTemplates: [{ type: ContentChildren, args: [TdDynamicFormsErrorTemplate,] }],
    elements: [{ type: Input, args: ['elements',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/** @type {?} */
const TD_DYNAMIC_FORMS = [
    TdDynamicFormsComponent,
    TdDynamicElementComponent,
    TdDynamicElementDirective,
    TdDynamicFormsErrorTemplate,
];
/** @type {?} */
const TD_DYNAMIC_FORMS_ENTRY_COMPONENTS = [
    TdDynamicInputComponent,
    TdDynamicFileInputComponent,
    TdDynamicTextareaComponent,
    TdDynamicSlideToggleComponent,
    TdDynamicCheckboxComponent,
    TdDynamicSliderComponent,
    TdDynamicSelectComponent,
    TdDynamicDatepickerComponent,
];
class CovalentDynamicFormsModule {
}
CovalentDynamicFormsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    TD_DYNAMIC_FORMS,
                    TD_DYNAMIC_FORMS_ENTRY_COMPONENTS,
                ],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatInputModule,
                    MatSelectModule,
                    MatCheckboxModule,
                    MatSliderModule,
                    MatSlideToggleModule,
                    MatIconModule,
                    MatButtonModule,
                    MatDatepickerModule,
                    CovalentCommonModule,
                    CovalentFileModule,
                ],
                exports: [
                    TD_DYNAMIC_FORMS,
                    TD_DYNAMIC_FORMS_ENTRY_COMPONENTS,
                ],
                providers: [
                    DYNAMIC_FORMS_PROVIDER,
                ],
                entryComponents: [TD_DYNAMIC_FORMS_ENTRY_COMPONENTS],
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */

export { CovalentDynamicFormsModule, TdDynamicFormsComponent, TdDynamicElementBase, _TdDynamicElementMixinBase, TdDynamicFormsErrorTemplate, TdDynamicElementDirective, TdDynamicElementComponent, DYNAMIC_FORMS_PROVIDER_FACTORY, TdDynamicType, TdDynamicElement, DYNAMIC_ELEMENT_NAME_REGEX, TdDynamicFormsService, DYNAMIC_FORMS_PROVIDER, TdDynamicTextareaComponent, TdDynamicSliderComponent, TdDynamicSlideToggleComponent, TdDynamicSelectComponent, TdDynamicInputComponent, TdDynamicFileInputComponent, TdDynamicDatepickerComponent, TdDynamicCheckboxComponent };
