// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-checkbox/dynamic-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
var TdDynamicCheckboxComponent = /** @class */ (function () {
    function TdDynamicCheckboxComponent() {
        this.label = '';
        this.name = '';
        this.hint = '';
        this.required = false;
    }
    TdDynamicCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-checkbox',
                    template: "<div class=\"td-dynamic-checkbox-wrapper\">\n  <mat-checkbox [formControl]=\"control\"\n                [name]=\"name\"\n                [required]=\"required\">\n    {{label}}\n  </mat-checkbox>\n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                    styles: [""]
                }] }
    ];
    return TdDynamicCheckboxComponent;
}());
export { TdDynamicCheckboxComponent };
if (false) {
    /** @type {?} */
    TdDynamicCheckboxComponent.prototype.control;
    /** @type {?} */
    TdDynamicCheckboxComponent.prototype.label;
    /** @type {?} */
    TdDynamicCheckboxComponent.prototype.name;
    /** @type {?} */
    TdDynamicCheckboxComponent.prototype.hint;
    /** @type {?} */
    TdDynamicCheckboxComponent.prototype.required;
}
