// @covalent/http/esm5/http.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { NgModule, Injector, InjectionToken } from '@angular/core';
import { HttpClientModule, HttpHandler } from '@angular/common/http';
import { InternalHttpService } from './actions/http.mixin';
import { TdHttpService, TdInterceptorBehaviorService } from './interceptors/http.service';
import { TdURLRegExpInterceptorMatcher } from './interceptors/url-regexp-interceptor-matcher.class';
/** @type {?} */
export var HTTP_CONFIG = new InjectionToken('HTTP_CONFIG');
/**
 * @param {?} handler
 * @param {?} injector
 * @param {?} config
 * @return {?}
 */
export function httpFactory(handler, injector, config) {
    return new TdHttpService(handler, new TdInterceptorBehaviorService(injector, new TdURLRegExpInterceptorMatcher(), config.interceptors));
}
/** @type {?} */
export var HTTP_INTERCEPTOR_PROVIDER = {
    provide: TdHttpService,
    useFactory: httpFactory,
    deps: [HttpHandler, Injector, HTTP_CONFIG],
};
var CovalentHttpModule = /** @class */ (function () {
    function CovalentHttpModule(_internalHttpService) {
        this._internalHttpService = _internalHttpService;
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    CovalentHttpModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        if (config === void 0) { config = { interceptors: [] }; }
        return {
            ngModule: CovalentHttpModule,
            providers: [{
                    provide: HTTP_CONFIG,
                    useValue: config,
                },
                HTTP_INTERCEPTOR_PROVIDER,
            ],
        };
    };
    CovalentHttpModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        HttpClientModule,
                    ],
                    providers: [
                        InternalHttpService,
                    ],
                },] }
    ];
    /** @nocollapse */
    CovalentHttpModule.ctorParameters = function () { return [
        { type: InternalHttpService }
    ]; };
    return CovalentHttpModule;
}());
export { CovalentHttpModule };
if (false) {
    /** @type {?} */
    CovalentHttpModule.prototype._internalHttpService;
}
