// @covalent/dynamic-forms/esm5/dynamic-elements/dynamic-slider/dynamic-slider.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Component, ChangeDetectorRef } from '@angular/core';
var TdDynamicSliderComponent = /** @class */ (function () {
    function TdDynamicSliderComponent(_changeDetectorRef) {
        this._changeDetectorRef = _changeDetectorRef;
        this.label = '';
        this.required = undefined;
        this.name = '';
        this.hint = '';
        this.min = undefined;
        this.max = undefined;
    }
    /**
     * @return {?}
     */
    TdDynamicSliderComponent.prototype._handleBlur = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () {
            _this._changeDetectorRef.markForCheck();
        });
    };
    TdDynamicSliderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'td-dynamic-slider',
                    template: "<div class=\"td-dynamic-slider-wrapper mat-form-field mat-form-field-can-float mat-form-field-should-float\"\n      [class.mat-focused]=\"slider._isActive\">\n  <span class=\"mat-form-field-label-wrapper\">\n    <label class=\"mat-form-field-label mat-primary td-slider-label\">\n      {{label}}\n      <span *ngIf=\"required && !control?.disabled\" class=\"mat-form-field-required-marker\">*</span>\n    </label>\n  </span>\n  <div class=\"td-dynamic-slider-field\">\n    <mat-slider #slider\n                class=\"td-dynamic-slider\"\n                [formControl]=\"control\"\n                [attr.name]=\"name\"\n                [min]=\"min\"\n                [max]=\"max\"\n                thumbLabel\n                tickInterval=\"auto\"\n                [required]=\"required\"\n                (blur)=\"_handleBlur()\">\n    </mat-slider>\n  </div>  \n  <span class=\"mat-hint td-dynamic-element-hint\">{{hint}}</span>\n</div>",
                    styles: [":host .td-dynamic-slider-wrapper{display:block}.td-dynamic-slider-field{position:relative;-ms-flex-direction:row;flex-direction:row;display:-ms-flexbox;display:flex;box-sizing:border-box}.td-dynamic-slider-field .td-dynamic-slider{-ms-flex:1;flex:1}"]
                }] }
    ];
    /** @nocollapse */
    TdDynamicSliderComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return TdDynamicSliderComponent;
}());
export { TdDynamicSliderComponent };
if (false) {
    /** @type {?} */
    TdDynamicSliderComponent.prototype.control;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.label;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.required;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.name;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.hint;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.min;
    /** @type {?} */
    TdDynamicSliderComponent.prototype.max;
    /** @type {?} */
    TdDynamicSliderComponent.prototype._changeDetectorRef;
}
